File: 2022/day16/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;

# Parse input file
sub parse_input {
    my ($text) = @_;
    my %valves;
    my %tunnels;

    for my $line (split /\n/, $text) {
        if ($line =~ /Valve (\w+) has flow rate=(\d+); tunnels? leads? to valves? (.+)/) {
            my ($name, $rate, $neighbors) = ($1, $2, $3);
            $valves{$name} = int($rate);
            $tunnels{$name} = [split /, /, $neighbors];
        }
    }

    return (\%valves, \%tunnels);
}

# BFS to compute shortest distances between relevant valves
sub compute_distances {
    my ($valves, $tunnels) = @_;

    # Only care about valves with flow > 0 plus starting valve AA
    my @relevant = ('AA');
    for my $v (keys %$valves) {
        push @relevant, $v if $valves->{$v} > 0;
    }

    my %distances;

    for my $start (@relevant) {
        $distances{$start} = {};
        my @queue = ([$start, 0]);
        my %visited = ($start => 1);

        while (@queue) {
            my ($curr, $dist) = @{shift @queue};

            # Check if current is a relevant valve (not start)
            if ($curr ne $start && exists $valves->{$curr} && $valves->{$curr} > 0) {
                $distances{$start}{$curr} = $dist;
            }
            # Special case for AA
            if ($curr ne $start && $curr eq 'AA') {
                $distances{$start}{$curr} = $dist;
            }

            for my $neighbor (@{$tunnels->{$curr}}) {
                unless ($visited{$neighbor}) {
                    $visited{$neighbor} = 1;
                    push @queue, [$neighbor, $dist + 1];
                }
            }
        }
    }

    return \%distances;
}

# Convert a set (array ref) to a string key for memoization
sub set_to_key {
    return join(',', sort @_);
}

# Part 1: Find maximum pressure release in 30 minutes
sub part1 {
    my ($text) = @_;
    my ($valves, $tunnels) = parse_input($text);
    my $distances = compute_distances($valves, $tunnels);

    # Get valuable valves (flow > 0)
    my @valuable = grep { $valves->{$_} > 0 } keys %$valves;
    my %valuable_set = map { $_ => 1 } @valuable;

    my %memo;

    my $dfs;
    $dfs = sub {
        my ($pos, $time_left, $opened_ref) = @_;

        return 0 if $time_left <= 0;

        my $key = "$pos:$time_left:" . set_to_key(@$opened_ref);
        return $memo{$key} if exists $memo{$key};

        my %opened = map { $_ => 1 } @$opened_ref;
        my $best = 0;

        for my $next_valve (@valuable) {
            next if $opened{$next_valve};

            # Time to move there and open it
            my $time_cost = $distances->{$pos}{$next_valve} + 1;
            if ($time_cost < $time_left) {
                my $new_time = $time_left - $time_cost;
                my $pressure = $valves->{$next_valve} * $new_time;
                my @new_opened = (@$opened_ref, $next_valve);
                my $result = $pressure + $dfs->($next_valve, $new_time, \@new_opened);
                $best = $result if $result > $best;
            }
        }

        $memo{$key} = $best;
        return $best;
    };

    return $dfs->('AA', 30, []);
}

# Part 2: Find maximum pressure with elephant helper (26 minutes each)
sub part2 {
    my ($text) = @_;
    my ($valves, $tunnels) = parse_input($text);
    my $distances = compute_distances($valves, $tunnels);

    # Get valuable valves (flow > 0)
    my @valuable = sort grep { $valves->{$_} > 0 } keys %$valves;
    my $n = scalar @valuable;

    # Compute max pressure for each subset of valves (represented as bitmask)
    sub max_pressure_for_mask {
        my ($mask, $valuable_ref, $valves, $distances) = @_;

        # Build subset from mask
        my @subset;
        for my $i (0 .. $#{$valuable_ref}) {
            push @subset, $valuable_ref->[$i] if $mask & (1 << $i);
        }

        return 0 unless @subset;

        my %subset_set = map { $_ => 1 } @subset;
        my %memo;

        my $dfs;
        $dfs = sub {
            my ($pos, $time_left, $opened_ref) = @_;

            return 0 if $time_left <= 0;

            my $key = "$pos:$time_left:" . set_to_key(@$opened_ref);
            return $memo{$key} if exists $memo{$key};

            my %opened = map { $_ => 1 } @$opened_ref;
            my $best = 0;

            for my $next_valve (@subset) {
                next if $opened{$next_valve};

                my $time_cost = $distances->{$pos}{$next_valve} + 1;
                if ($time_cost < $time_left) {
                    my $new_time = $time_left - $time_cost;
                    my $pressure = $valves->{$next_valve} * $new_time;
                    my @new_opened = (@$opened_ref, $next_valve);
                    my $result = $pressure + $dfs->($next_valve, $new_time, \@new_opened);
                    $best = $result if $result > $best;
                }
            }

            $memo{$key} = $best;
            return $best;
        };

        return $dfs->('AA', 26, []);
    }

    # Compute max scores for all masks
    my @max_scores;
    for my $mask (0 .. (1 << $n) - 1) {
        $max_scores[$mask] = max_pressure_for_mask($mask, \@valuable, $valves, $distances);
    }

    # Find best partition where you and elephant open disjoint sets
    my $best = 0;
    my $full_mask = (1 << $n) - 1;
    for my $mask (0 .. (1 << $n) - 1) {
        my $complement = $full_mask ^ $mask;
        if ($mask <= $complement) {
            my $total = $max_scores[$mask] + $max_scores[$complement];
            $best = $total if $total > $best;
        }
    }

    return $best;
}

# Main
my $script_dir = dirname(__FILE__);
my $input_file = "$script_dir/../input.txt";

open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
my $text = do { local $/; <$fh> };
close $fh;

print "Part 1: ", part1($text), "\n";
print "Part 2: ", part2($text), "\n";

File: 2022/day16/rust/solution.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::fs;

fn parse_input(text: &str) -> (HashMap<String, i32>, HashMap<String, Vec<String>>) {
    let mut valves = HashMap::new();
    let mut tunnels = HashMap::new();

    for line in text.trim().lines() {
        // Parse: "Valve XX has flow rate=N; tunnels lead to valves A, B, C"
        let parts: Vec<&str> = line.split_whitespace().collect();
        let name = parts[1].to_string();
        let rate: i32 = parts[4]
            .trim_end_matches(';')
            .split('=')
            .nth(1)
            .unwrap()
            .parse()
            .unwrap();

        let neighbors: Vec<String> = parts[9..]
            .iter()
            .map(|s| s.trim_end_matches(',').to_string())
            .collect();

        valves.insert(name.clone(), rate);
        tunnels.insert(name, neighbors);
    }

    (valves, tunnels)
}

fn compute_distances(
    valves: &HashMap<String, i32>,
    tunnels: &HashMap<String, Vec<String>>,
) -> HashMap<String, HashMap<String, i32>> {
    // Only care about valves with flow > 0 plus starting valve AA
    let mut relevant: Vec<String> = vec!["AA".to_string()];
    relevant.extend(valves.iter().filter(|(_, &v)| v > 0).map(|(k, _)| k.clone()));

    let mut distances: HashMap<String, HashMap<String, i32>> = HashMap::new();

    for start in &relevant {
        let mut dist_map = HashMap::new();
        let mut queue = VecDeque::new();
        let mut visited = HashSet::new();

        queue.push_back((start.clone(), 0));
        visited.insert(start.clone());

        while let Some((curr, dist)) = queue.pop_front() {
            if relevant.contains(&curr) && &curr != start {
                dist_map.insert(curr.clone(), dist);
            }

            if let Some(neighbors) = tunnels.get(&curr) {
                for neighbor in neighbors {
                    if !visited.contains(neighbor) {
                        visited.insert(neighbor.clone());
                        queue.push_back((neighbor.clone(), dist + 1));
                    }
                }
            }
        }

        distances.insert(start.clone(), dist_map);
    }

    distances
}

fn part1(text: &str) -> i32 {
    let (valves, tunnels) = parse_input(text);
    let distances = compute_distances(&valves, &tunnels);

    // Get valuable valves (flow > 0) and assign indices
    let valuable: Vec<String> = valves
        .iter()
        .filter(|(_, &v)| v > 0)
        .map(|(k, _)| k.clone())
        .collect();

    let n = valuable.len();
    let aa_idx = n; // Use n as index for AA

    // Build distance matrix (including AA)
    let mut dist_matrix = vec![vec![0i32; n + 1]; n + 1];
    for (i, v) in valuable.iter().enumerate() {
        for (j, u) in valuable.iter().enumerate() {
            if i != j {
                dist_matrix[i][j] = *distances.get(v).unwrap().get(u).unwrap();
            }
        }
        dist_matrix[i][aa_idx] = *distances.get(v).unwrap().get("AA").unwrap();
        dist_matrix[aa_idx][i] = *distances.get("AA").unwrap().get(v).unwrap();
    }

    let flow_rates: Vec<i32> = valuable.iter().map(|v| valves[v]).collect();

    // DFS with memoization using bitmask for opened valves
    let mut memo: HashMap<(usize, i32, u64), i32> = HashMap::new();

    fn dfs(
        pos: usize,
        time_left: i32,
        opened: u64,
        n: usize,
        aa_idx: usize,
        dist_matrix: &Vec<Vec<i32>>,
        flow_rates: &Vec<i32>,
        memo: &mut HashMap<(usize, i32, u64), i32>,
    ) -> i32 {
        if time_left <= 0 {
            return 0;
        }

        let key = (pos, time_left, opened);
        if let Some(&cached) = memo.get(&key) {
            return cached;
        }

        let mut best = 0;
        for next in 0..n {
            if opened & (1 << next) != 0 {
                continue;
            }

            let time_cost = dist_matrix[pos][next] + 1;
            if time_cost < time_left {
                let new_time = time_left - time_cost;
                let pressure = flow_rates[next] * new_time;
                let result = pressure
                    + dfs(
                        next,
                        new_time,
                        opened | (1 << next),
                        n,
                        aa_idx,
                        dist_matrix,
                        flow_rates,
                        memo,
                    );
                best = best.max(result);
            }
        }

        memo.insert(key, best);
        best
    }

    dfs(
        aa_idx,
        30,
        0,
        n,
        aa_idx,
        &dist_matrix,
        &flow_rates,
        &mut memo,
    )
}

fn part2(text: &str) -> i32 {
    let (valves, tunnels) = parse_input(text);
    let distances = compute_distances(&valves, &tunnels);

    // Get valuable valves (flow > 0) and assign indices
    let valuable: Vec<String> = valves
        .iter()
        .filter(|(_, &v)| v > 0)
        .map(|(k, _)| k.clone())
        .collect();

    let n = valuable.len();
    let aa_idx = n;

    // Build distance matrix
    let mut dist_matrix = vec![vec![0i32; n + 1]; n + 1];
    for (i, v) in valuable.iter().enumerate() {
        for (j, u) in valuable.iter().enumerate() {
            if i != j {
                dist_matrix[i][j] = *distances.get(v).unwrap().get(u).unwrap();
            }
        }
        dist_matrix[i][aa_idx] = *distances.get(v).unwrap().get("AA").unwrap();
        dist_matrix[aa_idx][i] = *distances.get("AA").unwrap().get(v).unwrap();
    }

    let flow_rates: Vec<i32> = valuable.iter().map(|v| valves[v]).collect();

    // Compute max pressure for each subset of valves
    fn max_pressure_for_subset(
        subset_mask: u64,
        n: usize,
        aa_idx: usize,
        dist_matrix: &Vec<Vec<i32>>,
        flow_rates: &Vec<i32>,
    ) -> i32 {
        let mut memo: HashMap<(usize, i32, u64), i32> = HashMap::new();

        fn dfs(
            pos: usize,
            time_left: i32,
            opened: u64,
            subset_mask: u64,
            n: usize,
            aa_idx: usize,
            dist_matrix: &Vec<Vec<i32>>,
            flow_rates: &Vec<i32>,
            memo: &mut HashMap<(usize, i32, u64), i32>,
        ) -> i32 {
            if time_left <= 0 {
                return 0;
            }

            let key = (pos, time_left, opened);
            if let Some(&cached) = memo.get(&key) {
                return cached;
            }

            let mut best = 0;
            for next in 0..n {
                // Only consider valves in the subset that aren't opened
                if subset_mask & (1 << next) == 0 || opened & (1 << next) != 0 {
                    continue;
                }

                let time_cost = dist_matrix[pos][next] + 1;
                if time_cost < time_left {
                    let new_time = time_left - time_cost;
                    let pressure = flow_rates[next] * new_time;
                    let result = pressure
                        + dfs(
                            next,
                            new_time,
                            opened | (1 << next),
                            subset_mask,
                            n,
                            aa_idx,
                            dist_matrix,
                            flow_rates,
                            memo,
                        );
                    best = best.max(result);
                }
            }

            memo.insert(key, best);
            best
        }

        dfs(
            aa_idx,
            26,
            0,
            subset_mask,
            n,
            aa_idx,
            dist_matrix,
            flow_rates,
            &mut memo,
        )
    }

    // Compute max scores for all subsets
    let total_subsets = 1u64 << n;
    let mut max_scores: Vec<i32> = vec![0; total_subsets as usize];

    for mask in 0..total_subsets {
        max_scores[mask as usize] =
            max_pressure_for_subset(mask, n, aa_idx, &dist_matrix, &flow_rates);
    }

    // Find best partition where you and elephant open disjoint sets
    let mut best = 0;
    let full_mask = total_subsets - 1;
    for mask in 0..total_subsets {
        let complement = full_mask ^ mask;
        if mask <= complement {
            let score = max_scores[mask as usize] + max_scores[complement as usize];
            best = best.max(score);
        }
    }

    best
}

fn main() {
    let input_path = std::env::current_dir()
        .unwrap()
        .join("../input.txt");
    let text = fs::read_to_string(input_path).expect("Failed to read input file");

    println!("Part 1: {}", part1(&text));
    println!("Part 2: {}", part2(&text));
}
